src/ossl/ecc.rs

```diff
@@ -0,0 +1,905 @@
+            i2d_ASN1_OCTET_STRING(
+                asn1_str,
+                &mut ($vec.as_mut_ptr()),
+            )
+        };
+        tmp.zeroize();
+        unsafe { ASN1_STRING_free(asn1_str); }
+        if pub_asn1_len as usize != $vec.len()
+        {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+    }};
+}
+
+macro_rules! param_octet_string_to_obj_attr {
+    ($params:expr; $name:expr; $obj:expr; $attr:expr) => {{
+        let mut vec = Vec::<u8>::new();
+        param_octet_string_to_vec!($params; $name; vec);
+        $obj.set_attr(attribute::from_bytes($attr, vec))?;
+    }};
+}
+
+// TODO taken from rsa.rs
+macro_rules! make_bn {
+    ($name:expr; $vin:expr; $vout:expr) => {{
+        let bn = unsafe {
+            BN_bin2bn(
+                $vin.as_ptr() as *mut u8,
+                $vin.len() as i32,
+                std::ptr::null_mut(),
+            )
+        };
+        if bn.is_null() {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        let mut param = unsafe {
+            OSSL_PARAM_construct_BN(
+                $name as *const u8 as *const i8,
+                std::ptr::null_mut(),
+                0,
+            )
+        };
+        /* calculate needed size */
+        unsafe {
+            OSSL_PARAM_set_BN(&mut param, bn);
+        }
+        $vout.resize(param.return_size, 0);
+        unsafe {
+            param.data = $vout.as_mut_ptr() as *mut std::os::raw::c_void;
+            param.data_size = $vout.len();
+            OSSL_PARAM_set_BN(&mut param, bn);
+        }
+        param
+    }};
+}
+
+macro_rules! make_bn_from_obj {
+    ($obj:expr; $id:expr; $name:expr; $vec:expr) => {{
+        let x = match $obj.get_attr_as_bytes($id) {
+            Ok(b) => b,
+            Err(_) => return err_rv!(CKR_DEVICE_ERROR),
+        };
+        make_bn!($name; x; $vec)
+    }};
+}
+
+macro_rules! param_bn_to_vec {
+    ($params:expr; $name:expr; $vec:expr) => {{
+        let p = unsafe {
+            OSSL_PARAM_locate($params, $name as *const u8 as *const i8)
+        };
+        if p.is_null() {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        let mut bn: *mut BIGNUM = std::ptr::null_mut();
+        if unsafe { OSSL_PARAM_get_BN(p, &mut bn) } != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        let big_num = BigNum::from_ptr(bn)?;
+        $vec.resize(bn_num_bytes(big_num.as_ptr()), 0);
+        if unsafe {
+            BN_bn2bin(
+                big_num.as_ptr(),
+                $vec.as_mut_ptr() as *mut std::os::raw::c_uchar,
+            ) as usize
+        } != $vec.len()
+        {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+    }};
+}
+
+macro_rules! param_bn_to_obj_attr {
+    ($params:expr; $name:expr; $obj:expr; $attr:expr) => {{
+        let mut vec = Vec::<u8>::new();
+        param_bn_to_vec!($params; $name; vec);
+        $obj.set_attr(attribute::from_bytes($attr, vec))?;
+    }};
+}
+
```

Comment:
This is a comment for all of the above:
Looks like a ton of macro rules here, any reason why these can't be just regular functions?
Debugging with macro_rules is extremely painful, so they should be used only when really needed, and for very small compact actions that will unlikely need to be debugged.